pkg/sync/sync_context.go

```diff
@@ -504,6 +514,44 @@ func (sc *syncContext) getSyncTasks() (_ syncTasks, successful bool) {
 		}
 	}
 
+	if sc.createNamespace && sc.namespace != "" {
+		isNamespaceCreationNeeded := true
+		//check if namespace exists as part of the resources
+		for _, res := range sc.resources {
+			if isNamespaceWithName(res.Target, sc.namespace) {
+				isNamespaceCreationNeeded = false
+				break
+			}
+		}
+		//check if namespace exists as part of the hooks
+		if isNamespaceCreationNeeded {
+			for _, res := range sc.hooks {
+				if isNamespaceWithName(res, sc.namespace) {
+					isNamespaceCreationNeeded = false
+					break
+				}
+			}
+		}
+
+		if isNamespaceCreationNeeded {
+			annotations := make(map[string]string)
+			annotations[common.AnnotationKeyHook] = common.SyncPhasePreSync
+			nsSpec := &v1.Namespace{TypeMeta: metav1.TypeMeta{APIVersion: "v1", Kind: kube.NamespaceKind}, ObjectMeta: metav1.ObjectMeta{Name: sc.namespace, Annotations: annotations}}
+			unstructuredObj, err := kube.ToUnstructured(nsSpec)
+			if err == nil {
+				liveObj, err := sc.kubectl.GetResource(sc.config, unstructuredObj.GroupVersionKind(), unstructuredObj.GetName(), "")
```

Comment:
Please add `errors.IsNotFound` check. If namespace is not found then we can just set liveObj to nil to indicate that namespace still have to be created.